### PR TITLE
Add RegistrationUrlBuilder

### PR DESCRIPTION
**Pull request by joelverhagen**
Progress on https://github.com/NuGet/NuGetGallery/issues/7736.
Depends on https://github.com/NuGet/NuGet.Services.Metadata/pull/708.

This will be used when building and updating entities with interlinking URLs.

### CODE REVIEW

File: src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            switch (hive)
+            {
+                case HiveType.Legacy:
+                    return _legacyBaseUrl;
+                case HiveType.Gzipped:
+                    return _gzippedBaseUrl;
+                case HiveType.SemVer2:
+                    return _semver2BaseUrl;
+                default:
+                    throw new NotImplementedException($"The hive type '{hive}' does not have configured base URL.");
```
Comment (xavierdecoster) [inline src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs]: nit: `The hive type '{hive}' does not have *a* configured base URL.`

File: src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        private static string GetPageFragment(NuGetVersion lower, NuGetVersion upper)
+        {
+            return
+                $"page/" +
+                $"{Uri.EscapeUriString(lower.ToNormalizedString().ToLowerInvariant())}/" +
+                $"{Uri.EscapeUriString(upper.ToNormalizedString().ToLowerInvariant())}";
+        }
+
+        public string GetLeafPath(string id, NuGetVersion version)
+        {
```
Comment (xavierdecoster) [inline src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs]: Null checks?

File: src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            return url.Substring(baseUrl.Length);
+        }
+
+        public string GetPageUrl(HiveType hive, string id, NuGetVersion lower, NuGetVersion upper)
+        {
+            return GetBaseUrl(hive) + GetPagePath(id, lower, upper);
+        }
+
+        public string GetPagePath(string id, NuGetVersion lower, NuGetVersion upper)
+        {
```
Comment (xavierdecoster) [inline src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs]: Null checks

File: src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            {
+                return null;
+            }
+
+            var path = ConvertToPath(fromHive, url);
+            return GetBaseUrl(toHive) + path;
+        }
+
+        public string ConvertToPath(HiveType hive, string url)
+        {
```
Comment (xavierdecoster) [inline src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs]: Null checks

File: src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+
+            _legacyBaseUrl = EnsureTrailingSlash(_options.Value.LegacyBaseUrl);
+            _gzippedBaseUrl = EnsureTrailingSlash(_options.Value.GzippedBaseUrl);
+            _semver2BaseUrl = EnsureTrailingSlash(_options.Value.SemVer2BaseUrl);
+        }
+
+        public string GetIndexPath(string id)
+        {
```
Comment (xavierdecoster) [inline src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs]: Null check

File: src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs (hunk 2 above)
Comment (joelverhagen) [reply src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs]: In general I don't put null checks are all the levels. This is a low level component so I thought we wouldn't need to null check here. If you feel strongly I can put them in.

File: src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs (hunk 1 above)
Comment (joelverhagen) [reply src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs]: Oh nice catch.

File: src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs (hunk 2 above)
Comment (xavierdecoster) [reply src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs]: Even though low level, it's a public API (from technical point of view). 
I personally would just add them to the places I highlighted, as the other public methods are calling into these and would be implicitly covered. Your call :)

Comment (joelverhagen) [reply src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs]: Okay, sounds good. I'll make that change.

Comment (joelverhagen) [reply src/NuGet.Jobs.Catalog2Registration/Schema/RegistrationUrlBuilder.cs]: Done.

**Pull request merged**